webgoat-lessons/jwt/src/main/java/org/owasp/webgoat/jwt/JWTFinalEndpoint.java

```diff
@@ -0,0 +1,195 @@
+			}
+		} catch (Exception e) {
+			encodedHeader="";
+			encodedPayload="signature type not supported in this tool, try jwt.io";
+			encodedSignature = "";
+		}
+		String result = "{\"encodedHeader\":\""+encodedHeader+"\",\"encodedPayload\":\""+encodedPayload+"\",\"encodedSignature\":\""+encodedSignature+"\"}";
+		//System.out.println(result);
+		return result;
+    }
+       
+    private byte[] getJWTSignature(String jsonHeader, String encodedHeader, String encodedPayload, String jsonSecret) throws NoSuchAlgorithmException, InvalidKeyException {
+    	String message = encodedHeader+"."+encodedPayload;
+  	  	String algorithm = "HmacSHA256";
+    	if (jsonHeader.equals("HS512")) {
+    		algorithm = "HmacSHA512";
+    	} 
+    	Mac macInstance = Mac.getInstance(algorithm);
+        SecretKeySpec secret_key = new SecretKeySpec(TextCodec.BASE64.decode(jsonSecret), algorithm);
+        macInstance.init(secret_key);
+
+        return macInstance.doFinal(message.getBytes(StandardCharsets.UTF_8));
+    }
+    
+    @PostMapping(path="/JWT/decode",produces=MediaType.TEXT_HTML_VALUE)
+    @ResponseBody
+    public String decode(@RequestParam("jwtToken") String jwtToken) throws NoSuchAlgorithmException {		
+		try {
+			String encodedHeader = jwtToken.substring(0, jwtToken.indexOf("."));
+			String encodedPayload = jwtToken.substring(jwtToken.indexOf(".")+1, jwtToken.lastIndexOf("."));
+			String jsonHeader = TextCodec.BASE64URL.decodeToString(encodedHeader);
+			String jsonPayload = TextCodec.BASE64URL.decodeToString(encodedPayload);
+			return "{\"jsonHeader\":\""+jsonHeader.replace("\"", "\\\"")+"\",\"jsonPayload\":\""+jsonPayload.replace("\"", "\\\"").replace("\t","").replace("\r", "").replace("\n", "")+"\"}";
+		} catch (Exception e) {
+			return "{\"jsonHeader\":\"\",\"jsonPayload\":\"\"}";
+		}
+    }
+    
+    @PostMapping("/JWT/final/follow/{user}")
+    public @ResponseBody
+    String follow(@PathVariable("user") String user) {
+        if ("Jerry".equals(user)) {
+            return "Following yourself seems redundant";
+        } else {
+            return "You are now following Tom";
+        }
+    }
+
+    @PostMapping("/JWT/final/delete")
+    public @ResponseBody
+    AttackResult resetVotes(@RequestParam("token") String token) {
+        if (StringUtils.isEmpty(token)) {
+            return failed(this).feedback("jwt-invalid-token").build();
+        } else {
+            try {
+                final String[] errorMessage = {null};
+                Jwt jwt = Jwts.parser().setSigningKeyResolver(new SigningKeyResolverAdapter() {
+                    @Override
+                    public byte[] resolveSigningKeyBytes(JwsHeader header, Claims claims) {
+                        final String kid = (String) header.get("kid");
+                        try (var connection = dataSource.getConnection()) {
+                            ResultSet rs = connection.createStatement().executeQuery("SELECT key FROM jwt_keys WHERE id = '" + kid + "'");
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *rs*. Execution paths that do not contain closure statements were detected. Ensure *rs* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).